transports/quic/src/connection.rs

```diff
@@ -0,0 +1,389 @@
+/// Tied to a specific endpoint.
+#[derive(Debug)]
+pub struct Connection {
+    /// Channel to the endpoint this connection belongs to.
+    endpoint_channel: EndpointChannel,
+    /// Pending message to be sent to the background task that is driving the endpoint.
+    pending_to_endpoint: Option<ToEndpoint>,
+    /// Events that the endpoint will send in destination to our local [`quinn_proto::Connection`].
+    /// Passed at initialization.
+    from_endpoint: mpsc::Receiver<quinn_proto::ConnectionEvent>,
+
+    /// The QUIC state machine for this specific connection.
+    connection: quinn_proto::Connection,
+    /// Identifier for this connection according to the endpoint. Used when sending messages to
+    /// the endpoint.
+    connection_id: quinn_proto::ConnectionHandle,
+    /// `Future` that triggers at the [`Instant`] that `self.connection.poll_timeout()` indicates.
+    next_timeout: Option<(Delay, Instant)>,
+}
+
+/// Error on the connection as a whole.
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum ConnectionError {
+    /// The task driving the endpoint has crashed.
+    #[error("Endpoint driver crashed")]
+    EndpointDriverCrashed,
+
+    /// Error in the inner state machine.
+    #[error("{0}")]
+    Quinn(#[from] quinn_proto::ConnectionError),
+}
+
+impl Connection {
+    /// Crate-internal function that builds a [`Connection`] from raw components.
+    ///
+    /// This function assumes that there exists a [`EndpointDriver`](super::endpoint::EndpointDriver)
+    /// that will process the messages sent to `EndpointChannel::to_endpoint` and send us messages
+    /// on `from_endpoint`.
+    ///
+    /// `connection_id` is used to identify the local connection in the messages sent to
+    /// `to_endpoint`.
+    ///
+    /// This function assumes that the [`quinn_proto::Connection`] is completely fresh and none of
+    /// its methods has ever been called. Failure to comply might lead to logic errors and panics.
+    pub fn from_quinn_connection(
+        endpoint_channel: EndpointChannel,
+        connection: quinn_proto::Connection,
+        connection_id: quinn_proto::ConnectionHandle,
+        from_endpoint: mpsc::Receiver<quinn_proto::ConnectionEvent>,
+    ) -> Self {
+        debug_assert!(!connection.is_closed());
+        Connection {
+            endpoint_channel,
+            pending_to_endpoint: None,
+            connection,
+            next_timeout: None,
+            from_endpoint,
+            connection_id,
+        }
+    }
+
+    /// The address that the local socket is bound to.
+    pub fn local_addr(&self) -> &SocketAddr {
+        self.endpoint_channel.socket_addr()
+    }
+
+    /// Returns the address of the node we're connected to.
+    pub fn remote_addr(&self) -> SocketAddr {
+        self.connection.remote_address()
+    }
+
+    /// Returns the ID of the node we're connected to.
+    pub fn remote_peer_id(&self) -> PeerId {
+        let session = self.connection.crypto_session();
```

Comment:
This function was only called once, when the upgrade on teh connection succeeded. Moved the code so that `ConnectionEstablished` already includes the `PeerId`: 963d3dd30fac8fe54b25e7e98e3041b809918b3a